File: 백준 9251 LCS -dp-/복습/Main.java
import java.util.Scanner;

public class Main {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String line1 = sc.next();
        String line2 = sc.next();
        int[][] lcs = new int[line1.length() + 1][line2.length() + 1];
        for (int i = 1; i <= line1.length(); i++) {
            for (int j = 1; j <= line2.length(); j++) {
                if (line1.charAt(i - 1) == line2.charAt(j - 1)) {
                    lcs[i][j] = lcs[i - 1][j - 1] + 1;
                } else {
                    lcs[i][j] = Math.max(lcs[i - 1][j], lcs[i][j - 1]);
                }
            }
        }
        System.out.println(lcs[line1.length()][line2.length()]);
    }
}

File: 백준 9251 LCS -dp-/복습/Main.kt
import java.util.*

fun main(args: Array<String>) {
    val sc = Scanner(System.`in`)
    val line1 = sc.next()
    val line2 = sc.next()
    val lcs = Array(line1.length + 1) { IntArray(line2.length + 1) }
    for (i in 1..line1.length) {
        for (j in 1..line2.length) {
            if (line1[i - 1] == line2[j - 1]) {
                lcs[i][j] = lcs[i - 1][j - 1] + 1
            } else {
                lcs[i][j] = Math.max(lcs[i - 1][j], lcs[i][j - 1])
            }
        }
    }
    println(lcs[line1.length][line2.length])
}